docs/honk.8
.\"
.\" Copyright (c) 2019 Ted Unangst
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate$
.Dt HONK 8
.Os
.Sh NAME
.Nm honk
.Nd honk administration
.Sh DESCRIPTION
The
.Nm
daemon processes messages from other federated servers.
This is the admin manual.
For user operation, see
.Xr honk 1 .
.Ss Setup
.Pp
Set up a TLS reverse proxy.
.Nm
can listen on TCP or unix sockets, but will not terminate TLS.
https is a required component for federation.
Also, http signature verification requires accurate time keeping.
.Pp
Make sure to pass the Host header, if necessary (as for nginx).
.Bd -literal -offset indent
proxy_set_header Host $http_host;
.Ed
.Pp
FastCGI can be used by prefixing the listen address with "fcgi:".
.Ss Build
Building
.Nm
requires a go compiler 1.21+ and libsqlite.
On
.Ox
this is the go and sqlite3 packages.
Other platforms may require additional development libraries or headers
to be installed.
Run make.
Please be patient.
Even on fast machines, building from source can take several seconds.
.Ss Options
The following options control where
.Nm
looks for data.
.Bl -tag -width datadirxdirx
.It Fl datadir Ar dir
The root data directory, where the database and other user data are stored.
This directory contains all user data that persists across upgrades.
Requires write access.
Defaults to ".".
.It Fl viewdir Ar dir
The root view directory, where html and other templates are stored.
The contents of this directory are generally replaced with each release.
Read only.
Defaults to ".".
.El
.Pp
The following options control log output.
Acceptable values include "stderr" (the default), "stdout", "null", "syslog",
or a file name.
syslog messages will be sent to the UUCP facility.
.Bl -tag -width errorlogxlogx
.It Fl errorlog Ar log
The error log.
Something bad has happened.
.It Fl infolog Ar log
The informative messages log.
Something has happened, but probably not too bad.
.It Fl debuglog Ar log
The debug log.
There's probably no reason to care.
.It Fl log Ar log
Set all three logs.
.El
.Ss Init
Run the
.Ic init
command.
This will create the database and ask four questions, as well as creating
the initial user.
See below about importing existing data.
.Ss Operation
Run honk.
.Ss Customization
The funzone contains fun flair that users may add to posts and profiles.
Add custom memes (stickers) to the
.Pa memes
data directory.
Image and video files are supported.
Add custom emus (emoji) to the
.Pa emus
data directory.
PNG and GIF files are supported.
.Pp
Site CSS may be overridden by creating a
.Pa views/local.css
file in the data directory.
Site JS may similarly be included by creating
.Pa views/local.js .
A restart is required after changes.
A site icon.png and favicon.ico will be served from the views directory
in the data directory, if present.
.Pp
Custom HTML messages may be added to select pages by using the
.Ic admin
command.
This interface is a little rough.
A restart is required after changes.
.Bl -tag -width tenletters
.It server
Displayed on the home page.
.It about
Displayed on the about page.
.It login
Displayed on the login form.
.It avatar colors
Four 32-bit hex colors (RGBA).
.El
.Pp
.Ss User Admin
New users can be added with the
.Ic adduser
command.
This is discouraged.
.Pp
Passwords may be reset with the
.Ic chpass Ar username
command.
.Pp
Users may be deleted with the
.Ic deluser Ar username
command.
.Pp
Follow and unfollow requests can be sent via command line with
.Ic follow Ar username Ar url
and
.Ic unfollow Ar username Ar url .
.Ss Maintenance
The database may grow large over time.
The
.Ic cleanup Op Ar days
command exists to purge old external data, by default 30 days.
This removes unreferenced, unsaved posts and attachments.
It does not remove any original content.
This will not immediately reduce the size of the database, but frees space
for future use.
A vacuum may be performed manually if necessary, but will require more time
and additional disk space.
.Pp
Backups may be performed by running
.Ic backup dirname .
Backups only include the minimal necessary information, such as user posts
and follower information, but not external posts.
.Pp
Sometimes servers simply disappear, resulting in many errors trying to deliver
undeliverable messages.
Running
.Ic unplug Ar hostname
will delete all subscriptions and pending deliveries.
.Ss Upgrade
Stop the old honk process.
Backup the database.
Perform the upgrade with the
.Ic upgrade
command.
Restart.
.Pp
The current version of the honk binary may be printed with the
.Ic version
command.
.Ss Security
.Nm
is not currently hardened against SSRF, server side request forgery.
Be mindful of what other services may be exposed via localhost or the
local network.
.Ss Development
Development mode may be enabled or disabled by running
.Ic devel Ar on|off .
In devel mode, secure cookies are disabled, TLS certs are not verified,
and templates are reloaded every request.
.Ss Import
Data may be imported and converted from other services using the
.Ic import
command.
Currently supports Honk, Mastodon, Twitter, and Instagram exported data.
Posts are imported and backdated to appear as old honks.
The Mastodon following list is imported, but must be refollowed.
.Pp
To prepare a Honk data archive, extract the export.zip file.
.Dl ./honk import username honk source-directory
.Pp
To prepare a Mastodon data archive, extract the archive-longhash.tar.gz file.
.Dl ./honk import username mastodon source-directory
.Pp
To prepare a Twitter data archive, extract the twitter-longhash.zip file.
After unzipping the data archive, navigate to the tweet_media directory
and unzip any zip files contained within.
.Dl ./honk import username twitter source-directory
.Pp
To prepare an Instagram data archive, extract the igusername.zip file.
.Dl ./honk import username instagram source-directory
.Ss Export
User data may be exported to a zip archive using the
.Ic export
command.
This will export the user's outbox and inbox in ActivityPub json format,
along with associated media.
.Dl ./honk export username zipname
.Ss Advanced Options
Advanced configuration values may be set by running the
.Ic setconfig Ar key value
command.
For example, to increase the fast timeout value from 5 seconds to 10:
.Dl ./honk setconfig fasttimeout 10
.Pp
To support separate mentions without a subdomain,
e.g. @user@example.com and https://honk.example.com/u/user,
set config key 'masqname' to 'example.com'.
Route
.Pa /.well-known/webfinger
from the top domain to honk.
.Pp
Custom URL seperators (not "u" and "h") may be specified by adding
"usersep" and "honksep" options to the config table.
e.g. example.com/users/username/honk/somehonk instead of
example.com/u/username/h/somehonk.
.Sh FILES
.Nm
files are split between the data directory and the view directory.
Both default to "." but may be specified by command line options.
.Pp
The data directory contains:
.Bl -tag -width views/local.css
.It Pa honk.db
The main database.
.It Pa blob.db
Media and attachment storage.
.It Pa emus
Custom emoji.
.It Pa memes
Stickers and such.
.It Pa views/local.js
Locally customized JS.
.It Pa views/local.css
Locally customized CSS.
.El
.Pp
The view directory contains:
.Bl -tag -width views
.It Pa views
HTML templates and CSS files.
.El
.Sh EXAMPLES
This series of commands creates a new database, sets a friendly
welcome message, and runs honk.
.Bd -literal -offset indent
honk-v98> make
honk-v98> ./honk -datadir ../honkdata init
username: puffy
password: OxychromaticBlowfishSwatDynamite
listen address: /var/www/honk.sock
server name: honk.example.com
honk-v98> ./honk -datadir ../honkdata admin
honk-v98> date; ./honk -log honk.log -datadir ../honkdata
.Ed
.Pp
The views directory includes a sample mastodon.css to change color scheme.
.Bd -literal -offset indent
honk-v98> mkdir ../honkdata/views
honk-v98> cp views/mastodon.css ../honkdata/views/local.css
.Ed
.Pp
Upgrade to the next version.
Clean things up a bit.
.Bd -literal -offset indent
datadir> cp honk.db backup.db
datadir> cd ../honk-v99
honk-v99> make
honk-v99> ./honk -datadir ../honkdata upgrade
honk-v99> ./honk -datadir ../honkdata cleanup
honk-v99> date; ./honk -log honk.log -datadir ../honkdata
.Ed
.Sh ENVIRONMENT
Image processing and scaling requires considerable memory.
It is recommended to adjust the datasize ulimit to at least 1GB.
.Sh SEE ALSO
.Xr intro 1 ,
.Xr honk 1
.Sh CAVEATS
There's no online upgrade capability.
Upgrades may result in minutes of downtime.


honk.go
//
// Copyright (c) 2019 Ted Unangst <tedu@tedunangst.com>
//
// Permission to use, copy, modify, and distribute this software for any
// purpose with or without fee is hereby granted, provided that the above
// copyright notice and this permission notice appear in all copies.
//
// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
// ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
// ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
// OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

package main

import (
	"flag"
	"fmt"
	"golang.org/x/crypto/bcrypt"
	"html/template"
	golog "log"
	"log/syslog"
	notrand "math/rand"
	"os"
	"strconv"
	"strings"
	"time"

	"humungus.tedunangst.com/r/webs/httpsig"
	"humungus.tedunangst.com/r/webs/log"
)

var softwareVersion = "develop"

func init() {
	notrand.Seed(time.Now().Unix())
}

type WhatAbout struct {
	ID      int64
	Name    string
	Display string
	About   string
	HTAbout template.HTML
	Onts    []string
	Key     string
	URL     string
	Options UserOptions
	SecKey  httpsig.PrivateKey
}

type UserOptions struct {
	SkinnyCSS    bool   `json:",omitempty"`
	OmitImages   bool   `json:",omitempty"`
	MentionAll   bool   `json:",omitempty"`
	InlineQuotes bool   `json:",omitempty"`
	Avatar       string `json:",omitempty"`
	Banner       string `json:",omitempty"`
	MapLink      string `json:",omitempty"`
	Reaction     string `json:",omitempty"`
	MeCount      int64
	ChatCount    int64
}

type KeyInfo struct {
	keyname string
	seckey  httpsig.PrivateKey
}

const serverUID int64 = -2
const readyLuserOne int64 = 1

type Honk struct {
	ID       int64
	UserID   int64
	Username string
	What     string
	Honker   string
	Handle   string
	Handles  string
	Oonker   string
	Oondle   string
	XID      string
	RID      string
	Date     time.Time
	URL      string
	Noise    string
	Precis   string
	Format   string
	Convoy   string
	Audience []string
	Public   bool
	Whofore  int64
	Replies  []*Honk
	Flags    int64
	HTPrecis template.HTML
	HTML     template.HTML
	Style    string
	Open     string
	Donks    []*Donk
	Onts     []string
	Place    *Place
	Time     *Time
	Mentions []Mention
	Badonks  []Badonk
}

type Badonk struct {
	Who  string
	What string
}

type Chonk struct {
	ID     int64
	UserID int64
	XID    string
	Who    string
	Target string
	Date   time.Time
	Noise  string
	Format string
	Donks  []*Donk
	Handle string
	HTML   template.HTML
}

type Chatter struct {
	Target string
	Chonks []*Chonk
}

type Mention struct {
	Who   string
	Where string
}

func (mention *Mention) IsPresent(noise string) bool {
	nick := strings.TrimLeft(mention.Who, "@")
	idx := strings.IndexByte(nick, '@')
	if idx != -1 {
		nick = nick[:idx]
	}
	return strings.Contains(noise, ">@"+nick) || strings.Contains(noise, "@<span>"+nick)
}

func OntIsPresent(ont, noise string) bool {
	ont = strings.ToLower(ont[1:])
	idx := strings.IndexByte(noise, '#')
	for idx >= 0 {
		if strings.HasPrefix(noise[idx:], "#<span>") {
			idx += 6
		}
		idx += 1
		if idx+len(ont) >= len(noise) {
			return false
		}
		test := noise[idx : idx+len(ont)]
		test = strings.ToLower(test)
		if test == ont {
			return true
		}
		newidx := strings.IndexByte(noise[idx:], '#')
		if newidx == -1 {
			return false
		}
		idx += newidx
	}
	return false
}

type OldRevision struct {
	Precis string
	Noise  string
}

const (
	flagIsAcked    = 1
	flagIsBonked   = 2
	flagIsSaved    = 4
	flagIsUntagged = 8
	flagIsReacted  = 16
)

func (honk *Honk) IsAcked() bool {
	return honk.Flags&flagIsAcked != 0
}

func (honk *Honk) IsBonked() bool {
	return honk.Flags&flagIsBonked != 0
}

func (honk *Honk) IsSaved() bool {
	return honk.Flags&flagIsSaved != 0
}

func (honk *Honk) IsUntagged() bool {
	return honk.Flags&flagIsUntagged != 0
}

func (honk *Honk) IsReacted() bool {
	return honk.Flags&flagIsReacted != 0
}

type Donk struct {
	FileID   int64
	XID      string
	Name     string
	Desc     string
	URL      string
	Media    string
	Local    bool
	External bool
}

type Place struct {
	Name      string
	Latitude  float64
	Longitude float64
	Url       string
}

type Duration int64

func (d Duration) String() string {
	s := time.Duration(d).String()
	if strings.HasSuffix(s, "m0s") {
		s = s[:len(s)-2]
	}
	if strings.HasSuffix(s, "h0m") {
		s = s[:len(s)-2]
	}
	return s
}

func parseDuration(s string) time.Duration {
	didx := strings.IndexByte(s, 'd')
	if didx != -1 {
		days, _ := strconv.ParseInt(s[:didx], 10, 0)
		dur, _ := time.ParseDuration(s[didx:])
		return dur + 24*time.Hour*time.Duration(days)
	}
	dur, _ := time.ParseDuration(s)
	return dur
}

type Time struct {
	StartTime time.Time
	EndTime   time.Time
	Duration  Duration
}

type Honker struct {
	ID     int64
	UserID int64
	Name   string
	XID    string
	Handle string
	Flavor string
	Combos []string
	Meta   HonkerMeta
}

type HonkerMeta struct {
	Notes string
}

type SomeThing struct {
	What  int
	XID   string
	Owner string
	Name  string
}

const (
	SomeNothing int = iota
	SomeActor
	SomeCollection
)

var serverName string
var serverPrefix string
var masqName string
var dataDir = "./data"
var viewDir = "./views"
var iconName = "icon.png"
var serverMsg template.HTML
var aboutMsg template.HTML
var loginMsg template.HTML

func ElaborateUnitTests() {
}

func unplugserver(hostname string) {
	db := opendatabase()
	xid := fmt.Sprintf("%%https://%s/%%", hostname)
	db.Exec("delete from honkers where xid like ? and flavor = 'dub'", xid)
	db.Exec("delete from doovers where rcpt like ?", xid)
}

func reexecArgs(cmd string) []string {
	args := []string{"-datadir", dataDir}
	args = append(args, log.Args()...)
	args = append(args, cmd)
	return args
}

var elog, ilog, dlog *golog.Logger

func main() {
	log.Init(log.Options{Progname: "honk", Facility: syslog.LOG_UUCP})
	elog = log.E
	ilog = log.I
	dlog = log.D

	flag.StringVar(&dataDir, "datadir", dataDir, "data directory")
	flag.StringVar(&viewDir, "viewdir", viewDir, "view directory")
	flag.Parse()

	args := flag.Args()
	cmd := "run"
	if len(args) > 0 {
		cmd = args[0]
		args = args[1:]
	}

	switch cmd {
	case "genhash":
		os.Stderr.WriteString("This command generates bcrypt hashes for use with the init command.\n")
		os.Stderr.WriteString("To store the hash string output to a shell variable, do this:\n")
		os.Stderr.WriteString("hash=\"$(./honk genhash)\"\n")
		pass, err := askpassword()
		if err != nil {
			elog.Fatalf("error: %s\n", err.Error())
		}
		hash, err := bcrypt.GenerateFromPassword([]byte(pass), 12)
		if err != nil {
			elog.Fatalf("error: %s\n", err.Error())
		}
		os.Stderr.WriteString("For a flag option, copy the following line:\n")
		os.Stderr.WriteString("--hash \"" + strings.ReplaceAll(string(hash), "$", "\\$") + "\"\n")
		fmt.Println(string(hash))
		os.Exit(0)
	case "init":
		flags := flag.NewFlagSet(cmd, flag.ExitOnError)
		var hash string
		var listen string
		flags.StringVar(&hash, "hash", "", "password hash (bcrypt)")
		flags.StringVar(&listen, "listen", "0.0.0.0:8080", "listen address")
		err := flags.Parse(args)
		if err != nil {
			elog.Fatalf("failed parsing flags: %s\n", err.Error())
		}
		args = flags.Args()
		if len(args) < 2 {
			fmt.Printf("usage: init [--hash <hash>] [--listen <host:port>] <username> <fqdn>\n")
			return
		}
		initdb(args[0], hash, args[1], listen)
	case "upgrade":
		upgradedb()
	case "version":
		fmt.Println(softwareVersion)
		os.Exit(0)
	}

	db := opendatabase()
	dbversion := 0
	getconfig("dbversion", &dbversion)
	if dbversion != myVersion {
		elog.Fatal("incorrect database version. run upgrade.")
	}
	getconfig("servermsg", &serverMsg)
	getconfig("aboutmsg", &aboutMsg)
	getconfig("loginmsg", &loginMsg)
	getconfig("servername", &serverName)
	getconfig("masqname", &masqName)
	if masqName == "" {
		masqName = serverName
	}
	serverPrefix = fmt.Sprintf("https://%s/", serverName)
	getconfig("usersep", &userSep)
	getconfig("honksep", &honkSep)
	getconfig("devel", &develMode)
	getconfig("fasttimeout", &fastTimeout)
	getconfig("slowtimeout", &slowTimeout)
	getconfig("signgets", &signGets)
	prepareStatements(db)

	switch cmd {
	case "admin":
		adminscreen()
	case "import":
		if len(args) != 3 {
			elog.Fatal("usage: import <username> (mastodon|twitter) <srcdir>\n")
		}
		importMain(args[0], args[1], args[2])
	case "devel":
		if len(args) != 1 {
			elog.Fatal("usage: devel (on|off)")
		}
		switch args[0] {
		case "on":
			setconfig("devel", 1)
		case "off":
			setconfig("devel", 0)
		default:
			elog.Fatal("usage: devel (on|off)")
		}
	case "setconfig":
		if len(args) != 2 {
			elog.Fatal("usage: setconfig <key> <val>\n")
		}
		var val interface{}
		var err error
		if val, err = strconv.Atoi(args[1]); err != nil {
			val = args[1]
		}
		setconfig(args[0], val)
	case "adduser":
		flags := flag.NewFlagSet(cmd, flag.ExitOnError)
		var hash string
		flags.StringVar(&hash, "hash", "", "password hash (bcrypt)")
		err := flags.Parse(args)
		if err != nil {
			elog.Fatalf("failed parsing flags: %s\n", err.Error())
		}
		args = flags.Args()
		if len(args) < 1 {
			fmt.Printf("usage: honk init [--hash <hash>] <username>\n")
			return
		}
		adduser(args[0], hash)
	case "deluser":
		if len(args) < 1 {
			fmt.Printf("usage: honk deluser <username>\n")
			return
		}
		deluser(args[0])
	case "chpass":
		if len(args) < 1 {
			fmt.Printf("usage: honk chpass <username>\n")
			return
		}
		chpass(args[0])
	case "follow":
		if len(args) < 2 {
			fmt.Printf("usage: honk follow <username> <url>\n")
			return
		}
		user, err := butwhatabout(args[0])
		if err != nil {
			fmt.Printf("user not found\n")
			return
		}
		var meta HonkerMeta
		mj, _ := jsonify(&meta)
		honkerid, err := savehonker(user, args[1], "", "presub", "", mj)
		if err != nil {
			fmt.Printf("had some trouble with that: %s\n", err)
			return
		}
		followyou(user, honkerid, true)
	case "unfollow":
		if len(args) < 2 {
			fmt.Printf("usage: honk unfollow username url\n")
			return
		}
		user, err := butwhatabout(args[0])
		if err != nil {
			fmt.Printf("user not found\n")
			return
		}
		row := db.QueryRow("select honkerid from honkers where xid = ? and userid = ? and flavor in ('sub')", args[1], user.ID)
		var honkerid int64
		err = row.Scan(&honkerid)
		if err != nil {
			fmt.Printf("sorry couldn't find them\n")
			return
		}
		unfollowyou(user, honkerid, true)
	case "cleanup":
		arg := "30"
		if len(args) > 1 {
			arg = args[1]
		}
		cleanupdb(arg)
	case "unplug":
		if len(args) < 1 {
			fmt.Printf("usage: honk unplug <servername>\n")
			return
		}
		name := args[0]
		unplugserver(name)
	case "backup":
		if len(args) < 1 {
			fmt.Printf("usage: honk backup <dirname>\n")
			return
		}
		name := args[0]
		svalbard(name)
	case "ping":
		if len(args) < 2 {
			fmt.Printf("usage: honk ping (from username) (to username or url)\n")
			return
		}
		name := args[0]
		targ := args[1]
		user, err := butwhatabout(name)
		if err != nil {
			elog.Printf("unknown user")
			return
		}
		ping(user, targ)
	case "run":
		serve()
	case "backend":
		backendServer()
	case "test":
		ElaborateUnitTests()
	default:
		elog.Fatal("unknown command")
	}
}
